src/SignalR/clients/ts/signalr/src/HttpConnection.ts

```diff
@@ -269,29 +269,34 @@ export class HttpConnection implements IConnection {
             return;
         }
 
+        const transportExceptions: any[] = [];
         const transports = negotiateResponse.availableTransports || [];
         for (const endpoint of transports) {
-            this.connectionState = ConnectionState.Connecting;
-            const transport = this.resolveTransport(endpoint, requestedTransport, requestedTransferFormat);
-            if (typeof transport === "number") {
-                this.transport = this.constructTransport(transport);
-                if (!negotiateResponse.connectionId) {
-                    negotiateResponse = await this.getNegotiationResponse(url);
-                    connectUrl = this.createConnectUrl(url, negotiateResponse.connectionId);
-                }
-                try {
+            try {
+                this.connectionState = ConnectionState.Connecting;
+                const transport = this.resolveTransport(endpoint, requestedTransport, requestedTransferFormat);
+                if (typeof transport === "number") {
+                    this.transport = this.constructTransport(transport);
+                    if (!negotiateResponse.connectionId) {
+                        negotiateResponse = await this.getNegotiationResponse(url);
+                        connectUrl = this.createConnectUrl(url, negotiateResponse.connectionId);
+                    }
                     await this.transport!.connect(connectUrl, requestedTransferFormat);
                     this.changeState(ConnectionState.Connecting, ConnectionState.Connected);
                     return;
-                } catch (ex) {
-                    this.logger.log(LogLevel.Error, `Failed to start the transport '${HttpTransportType[transport]}': ${ex}`);
-                    this.connectionState = ConnectionState.Disconnected;
-                    negotiateResponse.connectionId = undefined;
                 }
+            } catch (ex) {
+                this.logger.log(LogLevel.Error, `Failed to start the transport '${endpoint.transport}': ${ex}`);
+                this.connectionState = ConnectionState.Disconnected;
```

Comment:
Hmm, I think we should move this outside of the loop. As long as we are in the loop we are technically still connecting. Rather than flip this back and forth.

Reply:
Unrelated to this PR, if you feel strongly about it you can make another PR :D

Reply:
Just merge this lol

Reply:
The extra state transitions aren't much of a problem anyway since there are no awaits while in the Disconnected state meaning they can't be observed elsewhere.